### DISCUSSION TITLE
Impure assigner actions

### DISCUSSION THREAD

**Discussion by phuoccss** [statelyai/xstate#1321] [General] [answered]
Hi. Is it an anti pattern if the assigner function is impure?

I am trying to implement undo/redo functionality using immer patches and handle those patches using another machine.

```js
{
  actions: {
    "undoableAction": assign(ctx => {
      const [nextCtx, patches, inversePatches] = produceWithPatches(...)

      someService.send('type', { patches })
      // or
      ctx.someServiceRef.send('type', { patches })

      return nextCtx
    })
  }
}

```

**Comment by Andarist** [comment, marked as answer]
> Hi. Is it an anti pattern if the assigner function is impure?

Yes. `assign`s should be pure. You could use `pure` though:
```js
{
  actions: {
    "undoableAction": pure(ctx => {
      const [nextCtx, patches, inversePatches] = produceWithPatches(/* ... */)
      return [
        assign(nextCtx),
        send({ type: 'type', patches }, { to: ctx.someServiceRef })
      ]
    })
  }
}
```

BTW you might be interested in this PR: https://github.com/davidkpiano/xstate/pull/1318

**Reply by phuoccss** [reply]
Thanks. Didn't know `pure` can be used this way.

The PR is interesting. Just that I need those patches for another service. I'm not quite sure how to get those with the PR.
